File: Subsets/subsets.scala
import scala.collection.mutable
object Solution {
    def subsets(nums: Array[Int]): List[List[Int]] = {
        
        val res = mutable.ListBuffer[List[Int]]()
        val curState = mutable.Stack[Int]()
        
        def uniqueSubset(cur: Int): Unit = {
            if (cur == nums.length) {
                res += curState.toList
            } else {
                uniqueSubset(cur + 1)
                curState push nums(cur)
                uniqueSubset(cur + 1)
                curState.pop
            }
        }
        
        uniqueSubset(0)
        return res.toList
    }

    def subsetsCascade(nums: Array[Int]): List[List[Int]] = {
        
        val res = mutable.ListBuffer[List[Int]](List[Int]())
        
        for (num <- nums) {
            for (subRes <- res.toList) {
                res += List(num) ::: subRes
            }
        }
        return res.toList
    }

}

File: Subsets/subsets.py
class Solution:
	
	# Using backtracking
	def subsets(self, nums: List[int]) -> List[List[int]]:
        curList = []
        res = []
        def uniqueSubsets(curInd):
            if curInd == len(nums):
                res.append(list(curList))
                return
                
            uniqueSubsets(curInd + 1)
            curList.append(nums[curInd])
            uniqueSubsets(curInd + 1)
            curList.pop()
        
        uniqueSubsets(0)
        return res

	# Cascading
	def subsetsCascading(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
		# for each number in nums, add it to existing outputs
        for i in nums:
            res += [[i] + subRes for subRes in res]
        return res

